base/src/main/java/io/spine/base/Environment.java

```diff
@@ -21,37 +21,71 @@
 package io.spine.base;
 
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import io.spine.annotation.Internal;
 import io.spine.annotation.SPI;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
+import static com.google.common.base.Preconditions.checkState;
+import static io.spine.base.BaseEnvironmentType.PRODUCTION;
+import static io.spine.base.BaseEnvironmentType.TESTS;
+
 /**
  * Provides information about the environment (current platform used, etc.).
+ *
+ * <p><b>When extending, please note</b> that this class does not handle the situations when two
+ * or more {@linkplain EnvironmentType environment types} return {@code true} on the
+ * {@link EnvironmentType#enabled()}. As such, if two or more user-defined environment types
+ * think that they are currently enabled, <b>the behaviour of {@link #currentType()} is
+ * undefined.</b>
  */
 @SPI
-@SuppressWarnings("AccessOfSystemProperties") // OK as we need system properties for this class.
 public final class Environment {
 
-    private static final Environment INSTANCE = new Environment();
+    private static final ImmutableList<EnvironmentType> BASE_TYPES =
+            ImmutableList.of(TESTS, PRODUCTION);
```

Comment:
Instead of putting the items explicitly, let's use `values()` of the enum type and  `copyOf()`.